WooCommerce/Classes/ViewRelated/Coupons/Add and Edit Coupons/AddEditCoupon.swift

```diff
@@ -1,9 +1,92 @@
 import SwiftUI
 import Yosemite
 
+final class AddEditCouponHostingController: UIHostingController<AddEditCoupon> {
+
+    private let viewModel: AddEditCouponViewModel
+
+    init(viewModel: AddEditCouponViewModel) {
+        self.viewModel = viewModel
+        super.init(rootView: AddEditCoupon(viewModel))
+
+        // Needed because a `SwiftUI` cannot be dismissed when being presented by a UIHostingController
+        rootView.dismissHandler = { [weak self] in
+            self?.dismiss(animated: true)
+        }
+    }
+
+    required dynamic init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        handleSwipeBackGesture()
+    }
+}
+
+extension AddEditCouponHostingController {
+    override func shouldPopOnBackButton() -> Bool {
+        guard !viewModel.hasChangesMade else {
+            presentDiscardChangesActionSheet(onDiscard: { [weak self] in
+                self?.discardAndPop()
+            })
+            return false
+        }
+        return true
+    }
+
+    override func shouldPopOnSwipeBack() -> Bool {
+        return shouldPopOnBackButton()
+    }
+}
+
+extension AddEditCouponHostingController: UIAdaptivePresentationControllerDelegate {
+    func presentationControllerShouldDismiss(_ presentationController: UIPresentationController) -> Bool {
+        return !viewModel.hasChangesMade
+    }
+
+    func presentationControllerDidAttemptToDismiss(_ presentationController: UIPresentationController) {
+        presentDiscardChangesActionSheet(onDiscard: { [weak self] in
+            self?.discardAndDismiss()
+        })
```

Comment:
We can simplify this to trigger the alert and discard changes directly here since there's only one type of navigation to support for this view:

```suggestion
        UIAlertController.presentDiscardChangesActionSheet(viewController: self, onDiscard: { [weak self] in
            self?.dismiss(animated: true)
        })
```